File: benchmarks/livermore_canonical/k10_diff_predict.f
C     Kernel 10 -- Difference predictors (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical: PX(25,101), CX(25,101)
C
      PROGRAM K10
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION PX(25,101), CX(25,101)
C
C     Initialize PX in column-major order: J outer, I inner
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 12 J = 1, 101
          DO 10 I = 1, 25
              BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
              FUZZ = -FUZZ
              PX(I,J) = (BUZZ - FIZZ) * 0.1D0
   10     CONTINUE
   12 CONTINUE
C
C     Initialize CX
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 22 J = 1, 101
          DO 20 I = 1, 25
              BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
              FUZZ = -FUZZ
              CX(I,J) = (BUZZ - FIZZ) * 0.1D0
   20     CONTINUE
   22 CONTINUE
C
C     Kernel loop
C
      N = 101
      DO 200 REP = 1, 6000000
          DO 90 K = 1, N
              AR       =      CX(5,K)
              BR       = AR - PX(5,K)
              PX(5,K)  = AR
              CR       = BR - PX(6,K)
              PX(6,K)  = BR
              AR       = CR - PX(7,K)
              PX(7,K)  = CR
              BR       = AR - PX(8,K)
              PX(8,K)  = AR
              CR       = BR - PX(9,K)
              PX(9,K)  = BR
              AR       = CR - PX(10,K)
              PX(10,K) = CR
              BR       = AR - PX(11,K)
              PX(11,K) = AR
              CR       = BR - PX(12,K)
              PX(12,K) = BR
              PX(14,K) = CR - PX(13,K)
              PX(13,K) = CR
   90     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) PX(14,1)
  900 FORMAT('px(14,1) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k10_diff_predict.c
/* K10 -- Difference predictors (canonical, 1-based indexing)
   Canonical: PX(25,101), CX(25,101)
   Column-major: arr(i,j) at flat (j-1)*25 + i (1-based) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     101
#define NREPS 6000000

/* 1-based Fortran column-major: arr(i,j) with dims (25,101) */
#define PX(i,j) px[((j)-1)*25 + (i)]
#define CX(i,j) cx[((j)-1)*25 + (i)]

int main(void) {
    /* Max flat: (101-1)*25 + 25 = 2525, need [2526] */
    double px[2526], cx[2526];

    signel(&px[1], 25 * 101);
    signel(&cx[1], 25 * 101);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        for (long k = 1; k <= N; k++) {
            double ar, br, cr;
            ar       =      CX(5,k);
            br       = ar - PX(5,k);
            PX(5,k)  = ar;
            cr       = br - PX(6,k);
            PX(6,k)  = br;
            ar       = cr - PX(7,k);
            PX(7,k)  = cr;
            br       = ar - PX(8,k);
            PX(8,k)  = ar;
            cr       = br - PX(9,k);
            PX(9,k)  = br;
            ar       = cr - PX(10,k);
            PX(10,k) = cr;
            br       = ar - PX(11,k);
            PX(11,k) = ar;
            cr       = br - PX(12,k);
            PX(12,k) = br;
            PX(14,k) = cr - PX(13,k);
            PX(13,k) = cr;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("px(14,1) = %f\n", PX(14,1));
    return 0;
}
